op-node/rollup/async/asyncgossiper.go

```diff
@@ -0,0 +1,123 @@
+package async
+
+import (
+	"context"
+	"sync/atomic"
+
+	"github.com/ethereum-optimism/optimism/op-service/eth"
+)
+
+// AsyncGossiper is a component for gossiping a payload while running the derivation pipeline
+// the gossiping is done in a separate goroutine, so that the derivation pipeline can continue without waiting for the gossiping to finish
+// it also maintains the currently gossiping payload, so it can be checked against other found payloads
+type AsyncGossiper struct {
+	running atomic.Bool
+	// channel to add new payloads to gossip
+	set chan *eth.ExecutionPayload
+	// channel to request getting the currently gossiping payload
+	get chan chan *eth.ExecutionPayload
+	// channel to request clearing the currently gossiping payload
+	clear chan struct{}
+
+	hasPayload     atomic.Bool
+	currentPayload *eth.ExecutionPayload
+	net            Network
+}
+
+// To avoid import cycles, we define a new Network interface here
+// this interface is compatable with driver.Network
+type Network interface {
+	PublishL2Payload(ctx context.Context, payload *eth.ExecutionPayload) error
+}
+
+func NewAsyncGossiper(net Network) *AsyncGossiper {
+	return &AsyncGossiper{
+		running: atomic.Bool{},
+		set:     make(chan *eth.ExecutionPayload, 1),
+		get:     make(chan chan *eth.ExecutionPayload),
+		clear:   make(chan struct{}),
+
+		hasPayload:     atomic.Bool{},
+		currentPayload: nil,
+		net:            net,
+	}
+}
+
+// Gossip is an exposed, synchronous function to send a payload to be gossiped into the AsyncGossiper
+func (p *AsyncGossiper) Gossip(payload *eth.ExecutionPayload) {
+	// send the payload to the newPayloads channel. this will block until the payload can be sent
+	p.set <- payload
+}
+
+// Get is an exposed, synchronous function to get the currently gossiping payload from the async AsyncGossiper
+func (p *AsyncGossiper) Get() *eth.ExecutionPayload {
+	c := make(chan *eth.ExecutionPayload)
+	// send the channel as a request. this will block until sent
+	p.get <- c
+	// return the response payload. this will block until received
+	return <-c
+}
+
+// HasPayload is an exposed, synchronous function to check if the AsyncGossiper is currently holding a payload
+func (p *AsyncGossiper) HasPayload() bool {
+	return p.hasPayload.Load()
+}
+
+// Clear is an exposed, synchronous function to clear the currently gossiping payload from the async AsyncGossiper's state
+func (p *AsyncGossiper) Clear() {
+	// send the signal to the clearPayload channel. this will block until the payload can be sent
+	p.clear <- struct{}{}
```

Comment:
This comment doesn't match up with the concept of clearing. But also it only blocks until the command is "accepted" by the other thread - the method will return before the payload has actually been cleared.

Reply:
The comment is way too low level, it was meant to say that this `Clear` function will block until the `Start` goroutine can pick up the signal. I think the comment is more confusing than helpful so I've removed this and the others like it in favour of comments above the function being clearer.